test/TestMysteryBoxSales.ts

```diff
@@ -0,0 +1,701 @@
+    await enumerableNftContract.connect(user_1).setApprovalForAll(mbContract.address, true);
+
+    {
+      const parameter = { ...createBoxPara, sell_all: true };
+      await mbContract.connect(user_1).createBox.apply(null, Object.values(parameter));
+      const logs = await mbContract.queryFilter(mbContract.filters.CreationSuccess());
+      const result = last(logs)!.args;
+      sell_all_box_id = result.box_id;
+      openBoxParameters.box_id = sell_all_box_id;
+    }
+    {
+      const nftBalance = await enumerableNftContract.balanceOf(await user_1.getAddress());
+      // half of the NFT ids owned
+      for (let i = 0; i < nftBalance.toNumber(); i += 2) {
+        const nftId = await enumerableNftContract.tokenOfOwnerByIndex(await user_1.getAddress(), i);
+        not_sell_all_nft_id_list.push(nftId);
+      }
+      const parameter = { ...createBoxPara, sell_all: false, nft_id_list: not_sell_all_nft_id_list };
+      await mbContract.connect(user_1).createBox.apply(null, Object.values(parameter));
+      const logs = await mbContract.queryFilter(mbContract.filters.CreationSuccess());
+      const result = last(logs)!.args;
+      not_sell_all_box_id = result.box_id;
+    }
+    for (let i = 0; i < 256; i++) {
+      await advanceBlock();
+    }
+  });
+
+  beforeEach(async () => {
+    snapshotId = await takeSnapshot();
+    MockDate.set(Date.now());
+  });
+
+  afterEach(async () => {
+    await revertToSnapShot(snapshotId);
+    // We also need to reset the "real time" (as what we do for evm)
+    MockDate.reset();
+  });
+
+  it("Should getNftListForSale work", async () => {
+    let nft_id_list: BigNumber[] = [];
+    const nftBalance = await enumerableNftContract.balanceOf(await user_1.getAddress());
+    // half of the NFT ids owned
+    for (let i = 0; i < nftBalance.toNumber(); i++) {
+      const nftId = await enumerableNftContract.tokenOfOwnerByIndex(await user_1.getAddress(), i);
+      nft_id_list.push(nftId);
+    }
+    const nftList = await mbContract.getNftListForSale(sell_all_box_id, 0, nftBalance);
+    expect(nftList).to.eql(nft_id_list);
+    {
+      // buy some NFT(s)
+      const parameters = { ...openBoxParameters, amount: MaxNumberOfNFT };
+      const tx_parameters = {
+        value: txParameters.value.mul(parameters.amount),
+      };
+      await mbContract.connect(user_2).openBox.apply(null, Object.values(addTxParameters(parameters, tx_parameters)));
+    }
+    const newNftList = await mbContract.getNftListForSale(sell_all_box_id, 0, nftBalance);
+    expect(newNftList.length).to.be.eq(nftList.length - MaxNumberOfNFT);
+
+    const purchaseList = await mbContract.getPurchasedNft(sell_all_box_id, await user_2.getAddress());
+    expect(purchaseList.length).to.be.eq(MaxNumberOfNFT);
+    {
+      const mergedList = newNftList.concat(purchaseList);
+      const difference = differenceWith(nftList, mergedList, (nftValue, mergedValue) => nftValue.eq(mergedValue));
```

Comment:
Is there a `expect` statement missing? you declared a variable `difference` but its value is never used.